File: byteback-test/test-scala-2.13.8/src/main/scala/byteback/test/exceptions/PotentialNegativeArraySize.scala
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --nas -o %t.bpl
 */
package byteback.test.exceptions

import byteback.specification.Contract.assertion
import byteback.specification.Operators._

import byteback.specification.Contract.Behavior
import byteback.specification.Contract.Raise
import byteback.specification.Contract.Return

class PotentialNegativeArraySize {

  @Behavior
  def always(n: Int): Boolean = {
    true
  }

  @Behavior
  def n_is_negative(n: Int): Boolean = {
    lt(n, 0)
  }

  @Behavior
  def n_is_positive(n: Int): Boolean = {
    gte(n, 0)
  }

  @Raise(exception = classOf[NegativeArraySizeException], when = "n_is_negative")
  @Return(when = "n_is_positive")
  def potentialNegativeArray(n: Int): Unit = {
    val a = new Array[Int](n)
  }

  @Return
  def safeArrayInstantiation(n: Int): Unit = {
    if (n >= 0) {
      val a = new Array[Int](n)
    }
  }

  @Raise(exception = classOf[NegativeArraySizeException], when = "always")
  def incorrectArrayInstantiation(n: Int): Unit = {
    val a = new Array[Int](-1)
  }

  @Return
  def CatchNegativeArraySizeException_WrappingNegativeArrayInstantiation_AlwaysEntersCatch(): Unit = {
    try {
      val a = new Array[Int](-1)
    } catch {
      case exception: NegativeArraySizeException => return
    }
    assertion(false)
  }

}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 5 verified, 0 errors
 */

File: byteback-test/test-java-8/src/main/java/byteback/test/exceptions/PotentialNegativeArraySize.java
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --nas -o %t.bpl
 */
package byteback.test.exceptions;

import static byteback.specification.Contract.assertion;
import static byteback.specification.Operators.*;

import byteback.specification.Contract.Behavior;
import byteback.specification.Contract.Raise;
import byteback.specification.Contract.Return;

public class PotentialNegativeArraySize {

	@Behavior
	public boolean always(int n) {
		return true;
	}

	@Behavior
	public boolean n_is_negative(int n) {
		return lt(n, 0);
	}

	@Behavior
	public boolean n_is_positive(int n) {
		return gte(n, 0);
	}

	@Raise(exception = NegativeArraySizeException.class, when = "n_is_negative")
	@Return(when = "n_is_positive")
	public void potentialNegativeArray(int n) {
		final int[] a = new int[n];
	}

	@Return
	public void safeArrayInstantiation(int n) {
		if (n >= 0) {
			final int[] a = new int[n];
		}
	}

	@Raise(exception = NegativeArraySizeException.class, when = "always")
	public void incorrectArrayInstantiation(int n) {
		final int[] a = new int[-1];
	}

	public void CatchNegativeArraySizeException_WrappingNegativeArrayInstantiation_AlwaysEntersCatch() {
		try {
			final int[] a = new int[-1];
		} catch (final NegativeArraySizeException exception) {
			return;
		}
		assertion(false);
	}

}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 5 verified, 0 errors
 */
